test/integration/apiserver/flowcontrol/concurrency_util_test.go

```diff
@@ -0,0 +1,506 @@
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/admission/plugin/webhook/testcerts"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	genericfeatures "k8s.io/apiserver/pkg/features"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	webhookutil "k8s.io/apiserver/pkg/util/webhook"
+	"k8s.io/apiserver/plugin/pkg/authorizer/webhook"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	v1 "k8s.io/client-go/tools/clientcmd/api/v1"
+	featuregatetesting "k8s.io/component-base/featuregate/testing"
+	"k8s.io/kubernetes/cmd/kube-apiserver/app/options"
+	"k8s.io/kubernetes/pkg/controlplane"
+	"k8s.io/kubernetes/test/integration/framework"
+)
+
+const (
+	requestConcurrencyLimitMetricsName = "apiserver_flowcontrol_request_concurrency_limit"
+	requestExecutionSecondsSumName     = "apiserver_flowcontrol_request_execution_seconds_sum"
+	requestExecutionSecondsCountName   = "apiserver_flowcontrol_request_execution_seconds_count"
+	priorityLevelSeatUtilSumName       = "apiserver_flowcontrol_priority_level_seat_utilization_sum"
+	priorityLevelSeatUtilCountName     = "apiserver_flowcontrol_priority_level_seat_utilization_count"
+	fakeworkDuration                   = 200 * time.Millisecond
+	testWarmUpTime                     = 2 * time.Second
+	testTime                           = 10 * time.Second
+)
+
+func setupWithAuthorizer(t testing.TB, maxReadonlyRequestsInFlight, maxMutatingRequestsInFlight int, authz authorizer.Authorizer) (*rest.Config, framework.TearDownFunc) {
+	_, kubeConfig, tearDownFn := framework.StartTestServer(t, framework.TestServerSetup{
+		ModifyServerRunOptions: func(opts *options.ServerRunOptions) {
+			// Ensure all clients are allowed to send requests.
+			opts.Authorization.Modes = []string{"AlwaysAllow"}
+			opts.GenericServerRunOptions.MaxRequestsInFlight = maxReadonlyRequestsInFlight
+			opts.GenericServerRunOptions.MaxMutatingRequestsInFlight = maxMutatingRequestsInFlight
+		},
+		ModifyServerConfig: func(config *controlplane.Config) {
+			config.GenericConfig.Authorization.Authorizer = authz
+		},
+	})
+	return kubeConfig, tearDownFn
+}
+
+type SumAndCount struct {
+	Sum   float64
+	Count int
+}
+
+type plMetrics struct {
+	execSeconds    SumAndCount
+	seatUtil       SumAndCount
+	availableSeats int
+}
+
+// metricSnapshot maps from a priority level label to
+// a plMetrics struct containing APF metrics of interest
+type metricSnapshot map[string]plMetrics
+
+// Client request latency measurement
+type clientLatencyMeasurement struct {
+	SumAndCount
+	SumSq float64 // latency sum of squares
+	Mu    sync.Mutex
+}
+
+func (clm *clientLatencyMeasurement) reset() {
+	clm.Mu.Lock()
+	clm.Sum = 0
+	clm.Count = 0
+	clm.SumSq = 0
+	clm.Mu.Unlock()
+}
+
+func (clm *clientLatencyMeasurement) update(duration float64) {
+	clm.Mu.Lock()
+	clm.Count += 1
+	clm.Sum += duration
+	clm.SumSq += duration * duration
+	clm.Mu.Unlock()
+}
+
+func (clm *clientLatencyMeasurement) getStats() clientLatencyStats {
```

Comment:
shouldn't this lock?

Reply:
I think that, at least, `type clientLatencyMeasurement` should have some comment(s) about concurrency.  And `func getStats` should obey what those comments say or elaborate on its exception.

Reply:
The lack of locking here is OK because this func is called after the concurrent goroutines stop updating the same data.  But there should be commentary (as I suggested above) so that readers like @lavalamp are guided to see the validity of the code.

Reply:
I suspected as much, but I think there's no harm in locking here and it would prove that a stray write at least won't be only partially counted?

Reply:
Agree and added lock/unlock